extensions/todo/src/todo.ts

```diff
@@ -0,0 +1,82 @@
+import { SetStateAction } from 'react';
+
+export interface Todo {
+  id: string;
+  title: string;
+  isCompleted: boolean;
+  urgent: boolean;
+  important: boolean;
+  quick: boolean;
+}
+
+// state modification actions
+
+type Action = SetStateAction<Todo[]>;
+
+export function create_todo(todo: Todo): Action {
+  return (todos: Todo[]) => insert_todo_in_todo_list(todo, [...(todos || [])]);
+}
+
+export function edit_todo(index: number, patch: Partial<Todo>): Action {
+  return (todos: Todo[]) =>
+    insert_todo_in_todo_list(
+      { ...(todos?.[index] as Todo), ...patch }, // add
+      (todos ?? []).filter((_, i) => i !== index), // filter = copy + remove
+    );
+}
+
+export function has_text(text: string): (todo: Todo) => boolean {
+  const query = text.split(' ').map((q) => q.trim().toLowerCase());
+
+  function subquery_applies(subquery: string): (todo: Todo) => boolean {
+    const regex = /^(urgent|u|quick|q|important|i)[ ]=(true|false|t|f|yes|no|y|n)$/im;
```

Comment:
> `[ ]` 

That's a weird way to include a space? Is this intentional or are you trying to match `"[ ]"`